contracts/timealarms/src/alarms.rs

```diff
@@ -90,6 +99,60 @@ impl TimeAlarms {
             .is_some();
         Ok(AlarmsStatusResponse { remaining_alarms })
     }
+
+    fn try_add_old(
+        &self,
+        storage: &mut dyn Storage,
+        address: Addr,
+        time: TimeSeconds,
+    ) -> ContractResult<()> {
+        self.time_alarms
+            .add(storage, address, Timestamp::from_seconds(time))
+            .map_err(Into::into)
+    }
+}
+
+pub(super) struct InDelivery;
+
+impl InDelivery {
+    const ALARMS: Deque<'_, TimeAlarm> = Deque::new("alarms_in_delivery");
+
+    fn add(storage: &mut dyn Storage, addr: Addr, time: TimeSeconds) -> ContractResult<()> {
+        Self::ALARMS
+            .push_back(
+                storage,
+                &TimeAlarm {
+                    addr,
+                    time: Uint64::new(time),
+                },
+            )
+            .map_err(Into::into)
+    }
+
+    pub fn delivered(storage: &mut dyn Storage) -> ContractResult<()> {
+        Self::ALARMS
+            .pop_front(storage)
+            .map(|maybe_alarm: Option<TimeAlarm>| debug_assert!(maybe_alarm.is_some()))
+            .map_err(Into::into)
+    }
+
+    pub fn failed(storage: &mut dyn Storage) -> ContractResult<()> {
+        Self::ALARMS
+            .pop_front(storage)
+            .map_err(Into::into)
+            .and_then(|maybe_alarm: Option<TimeAlarm>| {
+                maybe_alarm.ok_or(ContractError::ReplyOnEmptyAlarmQueue)
+            })
+            .and_then(|TimeAlarm { addr, time }| {
+                TimeAlarms::new().try_add_old(storage, addr, time.u64())
+            })
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct TimeAlarm {
+    addr: Addr,
+    time: Uint64,
```

Comment:
refrain from using cosmwasm types so deep in our design

Reply:
Using it only because of it's guarantees for proper (de)serialization and mentioned limitations of JSON engines.